Feature/stereo

ZacharyTaylor opened the pull request:
Big change to the image undistort node.
-Switched to a new class for storing and modifying camera parameters, as using camera_info messages was fairly ugly and quite limiting.
-Added the ability to get output parameters from a ros topic or have them automatically set to good values.
-Added the ability to do rectifications
-Added the ability to convert images and scale them easily
-Added a second node stereo_info_node that sets up all of the camera info needed to do dense stereo reconstruction.
-Added a launchfile that uses stereo_info_node and 2 undistort nodes in combination with stereo_image_proc to generate stereo depthmaps.

ZacharyTaylor commented:
Switched to nodelets running inside nodes. This allows the creation of a new dense_stereo_node that performs the full stereo pipeline inside a single node. @helenol @raghavkhanna

raghavkhanna commented on include/image_undistort/undistorter.h:
```diff
@@ -36,8 +36,8 @@ class Undistorter {
+  cv::UMat map_y_;
 };
 }
 
```
I think this is OpenCV3 which may be the default in 16.04, however, can we keep the Mat version in a different branch for the people still on 14.04?

raghavkhanna commented:
Works great @ZacharyTaylor :).
Here are some screenshots of this in action. Works well on grass as well. Uses between 30 -60% of one CPU over 3 threads on the NUC i7 (without GPU processing). Thanks a lot.
![screenshot from 2016-11-17 00 58 46](https://cloud.githubusercontent.com/assets/5816698/20370928/8dbe6710-ac61-11e6-9aaf-ab223c504152.png)
![screenshot from 2016-11-17 00 59 37](https://cloud.githubusercontent.com/assets/5816698/20370929/8dc15696-ac61-11e6-99d9-7143d3f6a31f.png)

ZacharyTaylor commented:
@raghavkhanna I added some flags so hopefully it compiles as is under opencv 2 without issue now. I also changed the optimization flag from -O3 to -Ofast which made a big difference on my machine (speedup of 20%+)

ZacharyTaylor commented on src/camera_parameters.cpp:
```diff
@@ -0,0 +1,601 @@
+  // step
+  for (size_t i = 0; i < kFocalLengthEstimationAttempts; ++i) {
+    // get list of edge points to check
+    std::vector<Eigen::Vector2d, Eigen::aligned_allocator<Eigen::Vector2d>>
```
The undistorter needs these values as doubles as it makes no assumption about points coming from a 2d grid.

ZacharyTaylor commented on CMakeLists.txt:
```diff
@@ -1,30 +1,91 @@
+
 project(image_undistort)
 
-find_package(catkin_simple REQUIRED)
```
I followed the realsense driver as an example when setting up building nodelets which didn't use catkin simple and didn't switch back after I got things running
